File: creational-design-pattern/builder-pattern/builder.go
// Builder Pattern
// ================
// Category: Creational Design Pattern
//
// Intent:
//   Separate the construction of a complex object from its representation,
//   allowing step-by-step construction with a fluent, chainable API.
//
// When to use:
//   - When a struct has many fields, some optional
//   - When you want validation at build-time
//   - When you want readable, self-documenting construction code
//
// Key Participants:
//   - Builder: Accumulates configuration via chained method calls
//   - Product: The final struct
//   - Director (optional): Preset construction sequences
//
// Go note:
//   Go doesn't have constructors or default params. The builder pattern and
//   functional options pattern are the two main approaches for complex construction.

package main

import (
	"fmt"
	"strings"
)

// ---------------------------------------------------------------------------
// Product
// ---------------------------------------------------------------------------

// HttpRequest is the complex object we're building. Once built, treat as immutable.
type HttpRequest struct {
	Method          string
	URL             string
	Headers         map[string]string
	QueryParams     map[string]string
	Body            *string // nil means no body
	TimeoutSeconds  int
	Retries         int
	FollowRedirects bool
}

func (r *HttpRequest) String() string {
	var sb strings.Builder

	url := r.URL
	if len(r.QueryParams) > 0 {
		params := make([]string, 0, len(r.QueryParams))
		for k, v := range r.QueryParams {
			params = append(params, k+"="+v)
		}
		url += "?" + strings.Join(params, "&")
	}
	fmt.Fprintf(&sb, "%s %s\n", r.Method, url)

	for k, v := range r.Headers {
		fmt.Fprintf(&sb, "  %s: %s\n", k, v)
	}
	if r.Body != nil {
		preview := *r.Body
		if len(preview) > 80 {
			preview = preview[:80]
		}
		fmt.Fprintf(&sb, "  Body: %s...\n", preview)
	}
	fmt.Fprintf(&sb, "  Timeout: %ds | Retries: %d | Follow redirects: %v",
		r.TimeoutSeconds, r.Retries, r.FollowRedirects)

	return sb.String()
}

// ---------------------------------------------------------------------------
// Builder — fluent API with method chaining
// ---------------------------------------------------------------------------

type HttpRequestBuilder struct {
	method          string
	url             string
	headers         map[string]string
	queryParams     map[string]string
	body            *string
	timeout         int
	retries         int
	followRedirects bool
}

func NewHttpRequestBuilder() *HttpRequestBuilder {
	return &HttpRequestBuilder{
		method:          "GET",
		headers:         make(map[string]string),
		queryParams:     make(map[string]string),
		timeout:         30,
		followRedirects: true,
	}
}

// HTTP method shortcuts — each returns the builder for chaining
func (b *HttpRequestBuilder) Get(url string) *HttpRequestBuilder {
	b.method = "GET"
	b.url = url
	return b
}

func (b *HttpRequestBuilder) Post(url string) *HttpRequestBuilder {
	b.method = "POST"
	b.url = url
	return b
}

func (b *HttpRequestBuilder) Put(url string) *HttpRequestBuilder {
	b.method = "PUT"
	b.url = url
	return b
}

func (b *HttpRequestBuilder) Delete(url string) *HttpRequestBuilder {
	b.method = "DELETE"
	b.url = url
	return b
}

// Configuration methods
func (b *HttpRequestBuilder) Header(key, value string) *HttpRequestBuilder {
	b.headers[key] = value
	return b
}

func (b *HttpRequestBuilder) Query(key, value string) *HttpRequestBuilder {
	b.queryParams[key] = value
	return b
}

func (b *HttpRequestBuilder) Body(content string) *HttpRequestBuilder {
	b.body = &content
	return b
}

func (b *HttpRequestBuilder) Timeout(seconds int) *HttpRequestBuilder {
	b.timeout = seconds
	return b
}

func (b *HttpRequestBuilder) Retries(count int) *HttpRequestBuilder {
	b.retries = count
	return b
}

func (b *HttpRequestBuilder) NoFollowRedirects() *HttpRequestBuilder {
	b.followRedirects = false
	return b
}

// Build validates and produces the final HttpRequest.
func (b *HttpRequestBuilder) Build() (*HttpRequest, error) {
	if b.url == "" {
		return nil, fmt.Errorf("URL is required")
	}
	if (b.method == "POST" || b.method == "PUT") && b.body == nil {
		return nil, fmt.Errorf("%s requests should have a body", b.method)
	}

	// Copy maps so the builder can be reused without shared state
	headers := make(map[string]string, len(b.headers))
	for k, v := range b.headers {
		headers[k] = v
	}
	qp := make(map[string]string, len(b.queryParams))
	for k, v := range b.queryParams {
		qp[k] = v
	}

	return &HttpRequest{
		Method:          b.method,
		URL:             b.url,
		Headers:         headers,
		QueryParams:     qp,
		Body:            b.body,
		TimeoutSeconds:  b.timeout,
		Retries:         b.retries,
		FollowRedirects: b.followRedirects,
	}, nil
}

// ---------------------------------------------------------------------------
// Director (optional) — preset construction sequences
// ---------------------------------------------------------------------------

func JsonApiGet(url string) *HttpRequestBuilder {
	return NewHttpRequestBuilder().
		Get(url).
		Header("Accept", "application/json").
		Header("Content-Type", "application/json").
		Timeout(10).
		Retries(3)
}

func AuthenticatedPost(url, token, body string) *HttpRequestBuilder {
	return NewHttpRequestBuilder().
		Post(url).
		Header("Authorization", "Bearer "+token).
		Header("Content-Type", "application/json").
		Body(body).
		Timeout(30)
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
func main() {
	// Manual step-by-step building with chaining
	request, err := NewHttpRequestBuilder().
		Get("https://api.example.com/users").
		Header("Authorization", "Bearer abc123").
		Query("page", "1").
		Query("limit", "50").
		Timeout(15).
		Retries(2).
		Build()

	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Println("Manual build:")
	fmt.Println(request)

	// Director preset
	apiRequest, _ := JsonApiGet("https://api.example.com/products").
		Query("category", "electronics").
		Build()
	fmt.Println("\nDirector preset (JSON API GET):")
	fmt.Println(apiRequest)

	postRequest, _ := AuthenticatedPost(
		"https://api.example.com/orders",
		"secret-token",
		`{"item": "laptop", "qty": 1}`,
	).Build()
	fmt.Println("\nDirector preset (Authenticated POST):")
	fmt.Println(postRequest)
}

File: creational-design-pattern/builder-pattern/builder.rs
//! Builder Pattern
//! ================
//! Category: Creational Design Pattern
//!
//! Intent:
//!   Separate the construction of a complex object from its representation,
//!   allowing step-by-step construction with a fluent, chainable API.
//!
//! When to use:
//!   - When a struct has many fields, some optional
//!   - When you want compile-time or runtime validation before creation
//!   - When you want readable, self-documenting construction code
//!
//! Key Participants:
//!   - Builder: Accumulates configuration via chained method calls
//!   - Product: The final immutable struct
//!
//! Rust note:
//!   The builder pattern is extremely common in Rust because the language
//!   has no default parameters or method overloading. Libraries like
//!   `reqwest`, `tokio`, and `clap` all use builders extensively.

use std::collections::HashMap;

// ---------------------------------------------------------------------------
// Product — immutable once built
// ---------------------------------------------------------------------------
#[derive(Debug)]
struct HttpRequest {
    method: String,
    url: String,
    headers: HashMap<String, String>,
    query_params: HashMap<String, String>,
    body: Option<String>,
    timeout_seconds: u32,
    retries: u32,
    follow_redirects: bool,
}

impl std::fmt::Display for HttpRequest {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut url = self.url.clone();
        if !self.query_params.is_empty() {
            let qs: Vec<String> = self
                .query_params
                .iter()
                .map(|(k, v)| format!("{}={}", k, v))
                .collect();
            url = format!("{}?{}", url, qs.join("&"));
        }
        writeln!(f, "{} {}", self.method, url)?;
        for (k, v) in &self.headers {
            writeln!(f, "  {}: {}", k, v)?;
        }
        if let Some(ref body) = self.body {
            let preview: String = body.chars().take(80).collect();
            writeln!(f, "  Body: {}...", preview)?;
        }
        write!(
            f,
            "  Timeout: {}s | Retries: {} | Follow redirects: {}",
            self.timeout_seconds, self.retries, self.follow_redirects
        )
    }
}

// ---------------------------------------------------------------------------
// Builder — fluent API with method chaining
// ---------------------------------------------------------------------------
struct HttpRequestBuilder {
    method: String,
    url: String,
    headers: HashMap<String, String>,
    query_params: HashMap<String, String>,
    body: Option<String>,
    timeout: u32,
    retries: u32,
    follow_redirects: bool,
}

impl HttpRequestBuilder {
    fn new() -> Self {
        Self {
            method: "GET".to_string(),
            url: String::new(),
            headers: HashMap::new(),
            query_params: HashMap::new(),
            body: None,
            timeout: 30,
            retries: 0,
            follow_redirects: true,
        }
    }

    // HTTP method shortcuts — each consumes and returns self (move semantics)
    fn get(mut self, url: &str) -> Self {
        self.method = "GET".to_string();
        self.url = url.to_string();
        self
    }

    fn post(mut self, url: &str) -> Self {
        self.method = "POST".to_string();
        self.url = url.to_string();
        self
    }

    fn put(mut self, url: &str) -> Self {
        self.method = "PUT".to_string();
        self.url = url.to_string();
        self
    }

    fn delete(mut self, url: &str) -> Self {
        self.method = "DELETE".to_string();
        self.url = url.to_string();
        self
    }

    // Configuration methods
    fn header(mut self, key: &str, value: &str) -> Self {
        self.headers.insert(key.to_string(), value.to_string());
        self
    }

    fn query(mut self, key: &str, value: &str) -> Self {
        self.query_params.insert(key.to_string(), value.to_string());
        self
    }

    fn body(mut self, content: &str) -> Self {
        self.body = Some(content.to_string());
        self
    }

    fn timeout(mut self, seconds: u32) -> Self {
        self.timeout = seconds;
        self
    }

    fn retries(mut self, count: u32) -> Self {
        self.retries = count;
        self
    }

    fn no_follow_redirects(mut self) -> Self {
        self.follow_redirects = false;
        self
    }

    /// Validate and produce the immutable HttpRequest.
    fn build(self) -> Result<HttpRequest, String> {
        if self.url.is_empty() {
            return Err("URL is required".to_string());
        }
        if (self.method == "POST" || self.method == "PUT") && self.body.is_none() {
            return Err(format!("{} requests should have a body", self.method));
        }

        Ok(HttpRequest {
            method: self.method,
            url: self.url,
            headers: self.headers,
            query_params: self.query_params,
            body: self.body,
            timeout_seconds: self.timeout,
            retries: self.retries,
            follow_redirects: self.follow_redirects,
        })
    }
}

// ---------------------------------------------------------------------------
// Director (optional) — preset construction sequences
// ---------------------------------------------------------------------------
struct RequestDirector;

impl RequestDirector {
    fn json_api_get(url: &str) -> HttpRequestBuilder {
        HttpRequestBuilder::new()
            .get(url)
            .header("Accept", "application/json")
            .header("Content-Type", "application/json")
            .timeout(10)
            .retries(3)
    }

    fn authenticated_post(url: &str, token: &str, body: &str) -> HttpRequestBuilder {
        HttpRequestBuilder::new()
            .post(url)
            .header("Authorization", &format!("Bearer {}", token))
            .header("Content-Type", "application/json")
            .body(body)
            .timeout(30)
    }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
fn main() {
    // Manual step-by-step building with chaining
    let request = HttpRequestBuilder::new()
        .get("https://api.example.com/users")
        .header("Authorization", "Bearer abc123")
        .query("page", "1")
        .query("limit", "50")
        .timeout(15)
        .retries(2)
        .build()
        .expect("Failed to build request");

    println!("Manual build:");
    println!("{}\n", request);

    // Director preset
    let api_request = RequestDirector::json_api_get("https://api.example.com/products")
        .query("category", "electronics")
        .build()
        .expect("Failed to build request");

    println!("Director preset (JSON API GET):");
    println!("{}\n", api_request);

    let post_request = RequestDirector::authenticated_post(
        "https://api.example.com/orders",
        "secret-token",
        r#"{"item": "laptop", "qty": 1}"#,
    )
    .build()
    .expect("Failed to build request");

    println!("Director preset (Authenticated POST):");
    println!("{}", post_request);
}
